File: service/src/PingPong/Hub/GameHub.cs
using Microsoft.AspNetCore.SignalR;
using PingPong.Models;
using PingPong.Models.HubItems;
using PingPong.Models.PlayerIngameModel;
using PingPong.Models.PlayerPosition;

namespace PingPong.Hubs
{
    public class GameHub : Hub

    //Das 'RS' zeichen in Postman == Record Separator ( ASCII-Zeichen 30,hexadezimal 0x1E). HIER RAUSKOPIEREN --->  
    {

        public override async Task OnConnectedAsync()
        {
            await Clients.All.SendAsync("Receive Message", $" {Context.ConnectionId} has connected");
        }

        public async Task JoinLobby(Guid sessionId)
        {
            string groupName = sessionId.ToString();
            await Groups.AddToGroupAsync(Context.ConnectionId, groupName);
            await Clients.Group(groupName).SendAsync("ReceiveMessage", $"{Context.ConnectionId} has joined the lobby {groupName}");
        }

        public async Task LeaveLobby(Guid sessionId)
        {
            string groupName = sessionId.ToString();
            await Groups.RemoveFromGroupAsync(Context.ConnectionId, groupName);
            await Clients.Group(groupName).SendAsync("ReceiveMessage", $"{Context.ConnectionId} has left the lobby {groupName}");
        }

        public async Task DetectPlayerMovement(Guid sessionId, PlayerPosition playerPosition)
        {
            string groupName = sessionId.ToString();
            await Clients.Group(groupName).SendAsync("ReceivePlayerMovement", playerPosition.PlayerId, playerPosition.X, playerPosition.Y);
        }

        public async Task DetectPlayerSize(Guid sessionId, PlayerPosition playerSize)
        {
            string groupName = sessionId.ToString();
            Guid.Parse(playerSize.PlayerId.ToString());
            await Clients.Group(groupName).SendAsync("Received" + playerSize.X, playerSize.Y);
        }

        public async Task DetectPlayerScoreAndLife(Guid sessionId, PlayerScoreAndLife scoreAndLife)
        {
            string groupName = sessionId.ToString();
            await Clients.Group(groupName).SendAsync("ReceivedScoreAndLife", scoreAndLife.Score, scoreAndLife.Life);
        }

        public async Task DetectCurrentQuestion(Guid sessionId, QuestionItem item)
        {
            string groupName = sessionId.ToString();
            await Clients.Group(groupName).SendAsync("ReceivedCurrentQuestion", item.QuestionId);
        }

        public async Task DetectAnswerQuestion(Guid sessionId, AnswerQuestionItem item)
        {
            string groupName = sessionId.ToString();
            await Clients.Group(groupName).SendAsync("ReceivedAnsweredQuestion", item.Questionindex);
        }

        public async Task DetectBallMovement(Guid sessionId, BallPosition ballPosition)
        {
            string groupName = sessionId.ToString();
            await Clients.Group(groupName).SendAsync("ReceivedBallMovement", ballPosition.X, ballPosition.Y);
        }

        public async Task DetectQuestionBallMovement(Guid sessionId, BallPosition ballPosition)
        {
            string groupName = sessionId.ToString();
            await Clients.Group(groupName).SendAsync("ReceiveQuestiondBallMovement", ballPosition.X, ballPosition.Y);
        }
        public async Task DetectSpawnQuestionBall(Guid sessionId, BallPosition ballPosition)
        {
            string groupName = sessionId.ToString();
            await Clients.Group(groupName).SendAsync("ReceivedDetectSpawnQuestionBall", ballPosition.X, ballPosition.Y);
        }

        public async Task DetectBallSize(Guid sessionId, BallSize ballSize)
        {
            string groupName = sessionId.ToString();
            await Clients.Group(groupName).SendAsync("Received" + ballSize.Width, ballSize.Height);
        }
    }
}

File: client/src/api/gamehub.ts
import * as signalR from "@microsoft/signalr";
import { apiUrl } from "../utils/config";
import {
  AnswerQuestionItem,
  BallPosition,
  BallSize,
  PlayerPosition,
  PlayerScoreAndLife,
  QuestionItem,
} from "../utils/types";

export class GameHubClient {
  private connection: signalR.HubConnection;

  constructor() {
    this.connection = new signalR.HubConnectionBuilder()
      .withUrl(`${apiUrl}/hub`, {
        withCredentials: false, //  CORS policy
      })
      .configureLogging(signalR.LogLevel.Information)
      .build();
  }

  public async start(): Promise<void> {
    try {
      await this.connection.start();
      console.log("SignalR Connected.");
    } catch (err) {
      console.error("Error while starting connection: ", err);
      setTimeout(() => this.start(), 5000);
    }
  }

  public registerOnServerEvents(callbacks: {
    receiveMessage: (message: string) => void;
    receivePlayerMovement: (playerId: string, x: number, y: number) => void;
    receiveScoreAndLife: (score: number, life: number) => void;
    receivedCurrentQuestion: (questionId: string) => void;
    receiveBallMovement: (x: number, y: number) => void;
    receivedAnsweredQuestion: (questionindex: number) => void;
    receiveQuestionBallMovement: (x: number, y: number) => void;
    receivedDetectSpawnQuestionBall: (x: number, y: number) => void;
    receiveBallSize: (width: number, height: number) => void;
  }): void {
    this.connection.on("ReceiveMessage", callbacks.receiveMessage);
    this.connection.on(
      "ReceivePlayerMovement",
      callbacks.receivePlayerMovement
    );
    this.connection.on("ReceivedScoreAndLife", callbacks.receiveScoreAndLife);
    this.connection.on(
      "ReceivedCurrentQuestion",
      callbacks.receivedCurrentQuestion
    );
    this.connection.on(
      "ReceivedDetectSpawnQuestionBall",
      callbacks.receivedDetectSpawnQuestionBall
    );
    this.connection.on("ReceivedBallMovement", callbacks.receiveBallMovement);
    this.connection.on(
      "ReceiveQuestiondBallMovement",
      callbacks.receiveQuestionBallMovement
    );
    this.connection.on("ReceivedBallSize", callbacks.receiveBallSize);
    this.connection.on(
      "ReceivedAnsweredQuestion",
      callbacks.receivedAnsweredQuestion
    );
  }

  public async stop(): Promise<void> {
    try {
      await this.connection.stop();
      console.log("SignalR Disconnected.");
    } catch (err) {
      console.error("Error while stopping connection: ", err);
    }
  }

  public async joinLobby(sessionId: string): Promise<void> {
    await this.connection.invoke("JoinLobby", sessionId);
  }

  public async leaveLobby(sessionId: string): Promise<void> {
    await this.connection.invoke("LeaveLobby", sessionId);
  }

  public async detectPlayerMovement(
    sessionId: string,
    playerPosition: PlayerPosition
  ): Promise<void> {
    await this.connection.invoke(
      "DetectPlayerMovement",
      sessionId,
      playerPosition
    );
  }

  public async detectPlayerSize(
    sessionId: string,
    playerSize: PlayerPosition
  ): Promise<void> {
    await this.connection.invoke("DetectPlayerSize", sessionId, playerSize);
  }

  public async pushPlayerScoreAndLife(
    sessionId: string,
    scoreAndLife: PlayerScoreAndLife
  ): Promise<void> {
    await this.connection.invoke(
      "DetectPlayerScoreAndLife",
      sessionId,
      scoreAndLife
    );
  }

  public async pushCurrentQuestion(
    sessionId: string,
    item: QuestionItem
  ): Promise<void> {
    await this.connection.invoke(
      "DetectCurrentQuestion",
      sessionId,
      item
    );
  }

  public async detectBallMovement(
    sessionId: string,
    ballPosition: BallPosition
  ): Promise<void> {
    await this.connection.invoke("DetectBallMovement", sessionId, ballPosition);
  }

  public async pushQuestionBallMovement(
    sessionId: string,
    ballPosition: BallPosition
  ): Promise<void> {
    await this.connection.invoke(
      "DetectQuestionBallMovement",
      sessionId,
      ballPosition
    );
  }

  public async detectBallSize(
    sessionId: string,
    ballSize: BallSize
  ): Promise<void> {
    await this.connection.invoke("DetectBallSize", sessionId, ballSize);
  }

  public async pushAnsweredQuestion(
    sessionId: string,
    item: AnswerQuestionItem
  ): Promise<void> {
    await this.connection.invoke(
      "DetectAnswerQuestion",
      sessionId,
      item
    );
  }

  public async pushSpawnQuestionBall(
    sessionId: string,
    ballPosition: BallPosition
  ): Promise<void> {
    await this.connection.invoke(
      "DetectSpawnQuestionBall",
      sessionId,
      ballPosition
    );
  }
}
